epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityItem.kt

```diff
@@ -0,0 +1,207 @@
+        return height > 0 && width > 0
+    }
+
+    fun reset(newAdapterPosition: Int) {
+        fullyVisible = false
+        visible = false
+        focusedVisible = false
+        adapterPosition = newAdapterPosition
+        lastVisibleHeightNotified = NOT_NOTIFIED
+        lastVisibleWidthNotified = NOT_NOTIFIED
+        lastVisibilityNotified = NOT_NOTIFIED
+    }
+
+    fun handleVisible(epoxyHolder: EpoxyViewHolder, detachEvent: Boolean) {
+        val previousVisible = visible
+        visible = !detachEvent && isVisible()
+        if (visible != previousVisible) {
+            if (visible) {
+                epoxyHolder.visibilityStateChanged(VisibilityState.VISIBLE)
+            } else {
+                epoxyHolder.visibilityStateChanged(VisibilityState.INVISIBLE)
+            }
+        }
+    }
+
+    fun handleFocus(epoxyHolder: EpoxyViewHolder, detachEvent: Boolean) {
+        val previousFocusedVisible = focusedVisible
+        focusedVisible = !detachEvent && isInFocusVisible()
+        if (focusedVisible != previousFocusedVisible) {
+            if (focusedVisible) {
+                epoxyHolder.visibilityStateChanged(VisibilityState.FOCUSED_VISIBLE)
+            } else {
+                epoxyHolder.visibilityStateChanged(VisibilityState.UNFOCUSED_VISIBLE)
+            }
+        }
+    }
+
+    fun handlePartialImpressionVisible(
+        epoxyHolder: EpoxyViewHolder,
+        detachEvent: Boolean,
+        @IntRange(from = 0, to = 100) thresholdPercentage: Int
+    ) {
+        val previousPartiallyVisible = partiallyVisible
+        partiallyVisible = !detachEvent && isPartiallyVisible(thresholdPercentage)
+        if (partiallyVisible != previousPartiallyVisible) {
+            if (partiallyVisible) {
+                epoxyHolder.visibilityStateChanged(VisibilityState.PARTIAL_IMPRESSION_VISIBLE)
+            } else {
+                epoxyHolder.visibilityStateChanged(VisibilityState.PARTIAL_IMPRESSION_INVISIBLE)
+            }
+        }
+    }
+
+    fun handleFullImpressionVisible(epoxyHolder: EpoxyViewHolder, detachEvent: Boolean) {
+        val previousFullyVisible = fullyVisible
+        fullyVisible = !detachEvent && isFullyVisible()
+        if (fullyVisible != previousFullyVisible) {
+            if (fullyVisible) {
+                epoxyHolder.visibilityStateChanged(VisibilityState.FULL_IMPRESSION_VISIBLE)
+            }
+        }
+    }
+
+    fun handleChanged(epoxyHolder: EpoxyViewHolder, visibilityChangedEnabled: Boolean): Boolean {
+        var changed = false
+        if (visibleHeight != lastVisibleHeightNotified || visibleWidth != lastVisibleWidthNotified || viewVisibility != lastVisibilityNotified) {
+            if (visibilityChangedEnabled) {
+                if (viewVisibility == View.GONE) {
+                    epoxyHolder.visibilityChanged(0f, 0f, 0, 0)
+                } else {
+                    epoxyHolder.visibilityChanged(
+                        100f / height * visibleHeight,
```

Comment:
named parameters might make this a little more readable

Reply:
Pb is `EpoxyViewHolder` is still in Java, or do you meant to also covert it? I prefer to focus this change on visibility code.

Reply:
ah, my mistake. 🙏